### ISSUE TITLE
Optimize Graph.Node.removePredecessor()

### ISSUE THREAD

**Issue by nathanila** [cbeust/testng#1390]
### TestNG Version
6.11

### Actual behavior
Here's the code for `Graph.Node.removePredecessor()`.

```java
    public boolean removePredecessor(T o) {
      boolean result = false;

      T pred = m_predecessors.get(o);
      if (null != pred) {
        result = null != m_predecessors.remove(o);
        if (result) {
          ppp("  REMOVED PRED " + o + " FROM NODE " + m_object);
        }
        else {
          ppp("  FAILED TO REMOVE PRED " + o + " FROM NODE " + m_object);
        }
      }

      return result;
    }
```

This method calls `m_predecessors.get(o)`.  From the profile, I can tell it returns null most or all of the time since `m_predecessors.remove(o)` doesn't show up in the profile.  In fact, I instrumented the code and found that `m_predecessors` is empty 99.99998913091772% of the time.

`m_predecessors.get(o)` takes up all of the time for `Graph.Node.removePredecessor()`. This translates to half or more of the time for `TestRunner.initMethods()`.

`m_predecessors.get(o)` calls `HashMap.hash(o)`. `HashMap.hash(o)` takes all of the time.

### Expected behavior
Please change the code to the following:

```java
    public boolean removePredecessor(T o) {
      boolean result = false;

      if (m_predecessors.isEmpty())
         return(false);

      T pred = m_predecessors.get(o);
      //...
```

This avoids calling m_predecessors.get(o) and replaces it with a cheaper call which will short-circuit the method almost for every call.

I patched this change into my copy of TestNG.  This change eliminates the entire time as described above.

### Is the issue reproductible on runner?

- [ ] Shell
- [ ] Maven
- [ ] Gradle
- [ ] Ant
- [X] Eclipse
- [ ] IntelliJ
- [ ] NetBeans

### Test case sample
I'll see if I can post a picture of the profile.

**Comment by nathanila** [comment]
Here's a screenshot of the profile before making the change.

![image](https://cloud.githubusercontent.com/assets/1236594/24066199/f8ec9f24-0b35-11e7-8099-094a5f4f3829.png)

**Comment by krmahadevan** [comment]
@nathanila - Why not wrap this up with a pull request as well ? since you have taken so much time to pin-point the problem ?

**Comment by cbeust** [comment]
The fix looks harmless enough but I'm having a hard time believing that a `.get` on a hash map (which is constant time) is such a big deal that skipping it saves considerable amounts of time, even on a huge test suite.

Is there any way you could create (programmatically) such a huge suite to demonstrate that and include it in a pull request?

By the way, I really appreciate you going down to this level of details, I'm just curious to get to the bottom of this...

**Comment by juherr** [comment]
`m_predecessors.get(o)` should not take times but `o.hashCode()` may.
And as I remember, computing hashCode on a TestMethod is taking a lot of time.

**Comment by nathanila** [comment]
If you look at the profile `HashMap.get()` isn't taking the time but `HashMap.hash()` is.  Another proof is when I run with and without the code change, I see a noticable change in wall clock time from calling `TestNG.main()` to when `IMethodInterceptor,intercept()` is called.

I'll see if I can create a reproducing simplified test case.

As for pull requests, I don't know how yet.  If there is some documentation on how to do this in Eclipse, that would be very helpful.

**Comment by nathanila** [comment]
I am struggling to produce a reproducing simplified test case.  I can reproduce a situation where Graph.topologicalSort() takes a long time but not because of Graph.removeFromNodes().  Instead, Graph.isIndependent() and Graph.initializeIndependentNodes() are taking the time.  Any suggestions on how to get Graph.removeFromNodes() to execute more?

**Comment by steveprentice** [comment]
Likely dupe of #1710.

**Comment by juherr** [comment]
We expect the issue will be resolved by #1718.
